dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/State/TurnStateManager.cs

```diff
@@ -0,0 +1,149 @@
+    /// <typeparam name="TConversationState">Optional. Type of the conversation state object being persisted.</typeparam>
+    /// <typeparam name="TUserState">Optional. Type of the user state object being persisted.</typeparam>
+    /// <typeparam name="TTempState">Optional. Type of the temp state object being persisted.</typeparam>
+    public class TurnStateManager<TState, TConversationState, TUserState, TTempState> : ITurnStateManager<TState>
+        where TState : TurnState<TConversationState, TUserState, TTempState>, new()
+        where TConversationState : StateBase, new()
+        where TUserState : StateBase, new()
+        where TTempState : TempState, new()
+    {
+
+        public TurnStateManager() { }
+
+        /// <inheritdoc />
+        public async Task<TState> LoadStateAsync(IStorage? storage, ITurnContext turnContext)
+        {
+            try
+            {
+                // Compute state keys
+                Activity activity = turnContext.Activity;
+                string channelId = activity.ChannelId;
+                string botId = activity.Recipient.Id;
+                string conversationId = activity.Conversation.Id;
+                string userId = activity.From.Id;
+
+                Verify.ParamNotNull(activity, "TurnContext.Activity");
+                Verify.ParamNotNull(channelId, "TurnContext.Activity.ChannelId");
+                Verify.ParamNotNull(botId, "TurnContext.Activity.Recipient.Id");
+                Verify.ParamNotNull(conversationId, "TurnContext.Activity.Conversation.Id");
+                Verify.ParamNotNull(userId, "TurnContext.Activity.From.Id");
+
+                string conversationKey = $"{channelId}/${botId}/conversations/${conversationId}";
+                string userKey = $"{channelId}/${botId}/users/${userId}";
+
+                // read items from storage provider (if configured)
+                IDictionary<string, object> items;
+                if (storage != null)
+                {
+                    items = await storage.ReadAsync(new string[] { conversationKey, userKey });
+                }
+                else
+                {
+                    items = new Dictionary<string, object>();
+                }
+
+                TState state = new();
+                TUserState? userState = null;
+                TConversationState? conversationState = null;
+                TTempState? tempState = null;
+
+                if (items.TryGetValue(userKey, out object userStateValue))
+                {
+                    userState = userStateValue as TUserState;
+                }
+
+                if (items.TryGetValue(conversationKey, out object conversationStateValue))
+                {
+                    conversationState = conversationStateValue as TConversationState;
+                }
+
+                userState ??= new TUserState();
+                conversationState ??= new TConversationState();
+                tempState ??= new TTempState();
```

Comment:
Is `tempState` always null here since it's just temporary and not persisted?

Reply:
In addition, is it needed to ensure `tempState` has no storage key to prevent it from being persisted?

Reply:
temp state is supposed to be temporary and not persisted. Yes the storage key is what is keeping it from being persisted. And yes good call out, I fixed that.